src/fsm/mod.rs

```diff
@@ -0,0 +1,38 @@
+use core::cell::Cell;
+
+#[cfg(test)]
+mod test;
+
+pub enum FsmSignal<UserSignal> {
+    Entry,
+    Exit,
+    User(UserSignal),
+}
```

Comment:
Using an algebraic type allows us to mix built-in signals (`Entry`, `Exit`) with user signals and allows us to have strongly typed signal data since `UserSignal` can itself be an algebraic type.